### ISSUE TITLE
Provide a way to set environment variables when using cmake-presets

### ISSUE THREAD

**Issue by sterin** [microsoft/vscode-cmake-tools#1829]
For example, a preset that uses `vcpkg` requires a path to `vcpkg.cmake`, whose location may  vary across machines. As the [documentation ](https://github.com/microsoft/vscode-cmake-tools/blob/develop/docs/cmake-presets.md#vcpkg-integration)suggests, this can be done by using `$env{VCPKG_ROOT}`. 

Using the command line it is very easy to choose a vcpkg installation on a specific machine, or switch between vcpkg installations (to debug new versions of packages) by prepending the environment variable to the command line.

However, from inside vscode, there's no easy way to set or modify this variable. This makes it difficult to work with presets with CMake-Tools. 

Please add a setting for preset environment variables or enable the existing `cmake.environment` when using presets.

**Comment by andreeis** [comment]
A preset definition supports the "environment" keyword. Let us know if it is not working properly.

**Comment by sterin** [comment]
The `environment` keyword is used to set environment variables set by the preset to `CMakeLists.txt`. I would like a way to set the content of the environment passed to the preset. 

Say I have a configuration preset with:

```json
"cacheVariables": {
   "CMAKE_TOOLCHAIN_FILE": {
      "value": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake",
       "type": "FILEPATH"
    }
 },
```

From the command line I can easily switch between `vcpkg` installations by just changing the command line:

```shell
VCPKG_ROOT=/path/to/first/vcpkg/install cmake --preset ...
```
or 
```shell
VCPKG_ROOT=/path/to/second/vcpkg/install cmake --preset ...
```

But to do that from vscode, unless I am missing something very obvious, I need to close vscode, change the environment variable, and then start vscode again.

**Comment by esweet431** [comment]
You can set environment variables via the `environment` map in a Configure Preset. Example:

```
"environment": {
        "VCPKG_ROOT": "/path/to/vcpkg/install"
}
```

If you want to easily switch between vcpkg installs, I recommend defining 2 Configure Presets. The first Configure Preset can specify one vcpkg root. The second Configure Preset can inherit from the first Configure Preset and overwrite the VCPKG_ROOT env var. 

More info on the environment map is [here](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#configure-preset).

**Comment by bobbrow** [comment]
Because we don't invoke cmake with `--preset`, we actually resolve the `$env{}` variables while constructing the command line and it works to set `VCPKG_ROOT` in two hidden, base presets which are inherited by the two configure presets.  Does this not work on the command line?

```
        {
            "name": "base",
            "generator": "Visual Studio 16 2019",
            "architecture": {
                "value": "Win32",
                "strategy": "set"
            },
            "binaryDir": "${sourceDir}/out/build/${presetName}",
            "hidden": true
        },
        {
            "name": "base-env1",
            "inherits": "base",
            "environment": {
                "FOO": "BAR"
            },
            "hidden": true
        },
        {
            "name": "base-env2",
            "inherits": "base",
            "environment": {
                "FOO": "BAZ"
            },
            "hidden": true
        },
        {
            "name": "env1",
            "inherits": "base-env1",
            "cacheVariables": {
                "CMAKE_TOOLCHAIN_FILE": "$env{FOO}"
            }
        },
        {
            "name": "env2",
            "inherits": "base-env2",
            "cacheVariables": {
                "CMAKE_TOOLCHAIN_FILE": "$env{FOO}"
            }
        },

[cmakefileapi-driver] Configuring using preset
[cmakefileapi-driver] Invoking CMake C:\Program Files\CMake\bin\cmake.EXE with arguments ["-DCMAKE_TOOLCHAIN_FILE=BAZ","-Hd:/test/cmaketest","-Bd:/test/cmaketest/out/build/env2","-G","Visual Studio 16 2019","-A","Win32"]
```

**Comment by sterin** [comment]
Yes, this can be done this way, but I don't think it's a good idea in this scenario.

The file `CMakePresets.json` is usually source controlled. In my example it describes how to build the project using `vcpkg`, but it should not contain the exact location of `vcpkg` because it changes across machines, and it changes (a lot) when testing and debugging new vcpkg versions.

Changing `CMakePresets.json` each time is error prone because the changes can be easily commited with a bit of inattention, ruining other people's day.

**Comment by bobbrow** [comment]
What about CMakeUserPresets.json that has two entries that inherit a base preset from CMakePresets.json and set the environment differently in each one?  The user presets file is meant for such a case as this and is not meant to be committed to source control.

**Comment by esweet431** [comment]
Makes sense. For user-specific paths like VCPKG_ROOT, I would recommend defining two hidden Configure Presets in CMakeUserPresets.json, which is intended to save user-specific settings and be excluded from source control. 

In your case, you can have a Configure Preset in CMakePresets.json that defines all of your project-specific settings - let's name it `base`. Then you can define 2 Configure Presets in CMakeUserPresets.json that each inherit from `base` and specify VCPKG_ROOT via the `environment` map.

**Comment by sterin** [comment]
It is definitely possible to this. It is also possible to use a symbolic link to point to the vcpkg installation. I just think this overcomplicates things.

While `CMakePresets.json` is supposed to be the sole source of truth, the use of environment variables to set machine-specific values is standard practice, it even appears in CMake Tools' own [documentation](https://github.com/microsoft/vscode-cmake-tools/blob/develop/docs/cmake-presets.md#vcpkg-integration). 

What CMake Tools is missing is a way to set these values, or override these values, from inside vscode without going through hoops.

**Comment by globberwops** [comment]
This is currently a blocking issue for me and my team, and it is preventing us from switching from caches to presets.

Suppose you have say 3 different configure presets, 1-2 build presets for each configure preset, and maybe 2 different test presets.
Now, as a developer, you want to use a compiler launcher, a different compiler or compiler version, or something like @sterin  described.
You would have to write an awful amount of user presets just to be able to do that and since the build/test presets all refer to a configure preset, you would have to override them as well.

Not being able to modify the environment defeats the whole purpose of the `$penv{<variable-name>}` macro in presets.

**Comment by evakili** [comment]
It seems that in `cmakeDriver.ts` ([here](https://github.com/microsoft/vscode-cmake-tools/blob/90f81ad7fc46e7427d6f10dd1972e8151cd52081/src/drivers/cmakeDriver.ts#L200)) adding `environment` `configureEnvironment` to the env ignored when presets are enabled. IMO environments are not such things to be ignored because of `environment` prop in presets.

**Comment by elupus** [comment]
It's very confusing that these vscode configure options are ignored when using presets. They should have an effect when using presets too.

**Comment by HaoboGu** [comment]
Same issue here, any updates?

**Comment by gcampbell-msft** [comment]
@HaoboGu We have set this for 1.16 and we plan to investigate and design for this in the 1.16 release. Thanks for your interest.

**Comment by vipcxj** [comment]
Same issue here, any updates?

**Comment by gcampbell-msft** [comment]
@vipcxj This should be resolved by #3537 which is currently in pre-release. I am marking this as fixed and will close it when we release it to the official release! Thanks!
